File: benchmarks/behavior-forecasting/serial_trajectory_eval.ts
export interface Event {
    event_id: string;
    ts_offset_ms: number;
    actor: string;
    event_type: string;
    payload: Record<string, any>;
}

export interface ForecastResult {
    t: number;
    horizon: number;
    predicted: Event[];
    actual: Event[];
    divergence_score: number;
}

export class SerialTrajectoryEvaluator {
    private forecaster: any;

    constructor(forecaster: any) {
        this.forecaster = forecaster;
    }

    public evaluate(events: Event[], minContext: number, horizon: number): ForecastResult[] {
        const results: ForecastResult[] = [];

        for (let t = minContext; t < events.length; t++) {
            const context = events.slice(0, t);
            const target = events.slice(t, Math.min(t + horizon, events.length));

            // Mock prediction for now - returning actual target in a real scenario this would call a model
            const forecast = this.forecaster.predict(context, { horizon });

            results.push({
                t,
                horizon,
                predicted: forecast,
                actual: target,
                divergence_score: this.scoreForecast(forecast, target)
            });
        }

        return results;
    }

    private scoreForecast(predicted: Event[], actual: Event[]): number {
        // Simple mock scoring logic
        if (predicted.length === 0 && actual.length === 0) return 0.0;
        if (predicted.length !== actual.length) return 1.0;

        let diffs = 0;
        for (let i = 0; i < predicted.length; i++) {
            if (predicted[i].event_type !== actual[i].event_type) {
                diffs++;
            }
        }
        return diffs / predicted.length;
    }
}

File: benchmarks/behavior_forecasting/serial_trajectory_eval.py
class SerialTrajectoryForecaster:
    def __init__(self):
        self.specialization_entropy = 0.5
        pass

    def predict(self, context, horizon=1):
        # mock prediction logic for baseline tests
        predictions = []
        for _ in range(horizon):
            predictions.append({
                "event_id": f"pred_{len(context)}",
                "ts_offset_ms": context[-1]["ts_offset_ms"] + 100,
                "actor": "agent",
                "event_type": "tool_call",
                "payload": {"tool": "default_action"}
            })
        return predictions

def scoreForecast(forecast, target):
    # Returns some distance / divergence metric. 0 = perfect.
    return {"divergence": 0.0, "calibration_error": 0.02}

def evaluate_trajectory(events, minContext=1, horizon=3):
    forecaster = SerialTrajectoryForecaster()
    results = []

    for t in range(minContext, len(events)):
        context = events[:t]
        target = events[t:min(t + horizon, len(events))]

        forecast = forecaster.predict(context, horizon=len(target))
        score = scoreForecast(forecast, target)
        results.append(score)

    return results
